src/Sentry/Internal/BackgroundWorker.cs

```diff
@@ -59,20 +69,37 @@ public bool EnqueueEnvelope(Envelope envelope)
 
             if (Interlocked.Increment(ref _currentItems) > _maxItems)
             {
-                _ = Interlocked.Decrement(ref _currentItems);
+                Interlocked.Decrement(ref _currentItems);
                 _options.ClientReportRecorder.RecordDiscardedEvents(DiscardReason.QueueOverflow, envelope);
+                _options.LogDebug("Discarding envelope {0} because the queue is full.", envelope.TryGetEventId());
                 return false;
             }
 
+            _options.LogDebug("Enqueuing envelope {0}", envelope.TryGetEventId());
             _queue.Enqueue(envelope);
-            _ = _queuedEnvelopeSemaphore.Release();
+
+            if (process)
+            {
+                _queuedEnvelopeSemaphore.Release();
+            }
 
             return true;
         }
 
-        private async Task WorkerAsync()
+        /// <summary>
+        /// Processes the number of queued items specified.
+        /// Used only in testing, after calling <see cref="EnqueueEnvelope(Envelope, bool)"/>
+        /// when passing <c>process: false</c>.
+        /// </summary>
+        /// <param name="count">The number of items to process from the queue.</param>
+        public void ProcessQueuedItems(int count)
```

Comment:
If this is just for testing, could it be `internal` instead?

```suggestion
        internal void ProcessQueuedItems(int count)
```